### PR TITLE
[shape_estimation]change type

### PR DESCRIPTION
**Pull request by tkimura4**
## Description

add type of output by shape_estimation

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [ ] Read [pull request guidelines][pull-request-guidelines]
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[clang-tidy-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/ClangTidyGuideline/
[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[pull-request-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/PullRequestGuideline/
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by yukkysaito** [comment]
素直にconverter作っても良いかなという気もしますが、どう思います？

**Comment by tkimura4** [comment]
shape_estimation -> detection_by_trackerにおいて,
detection_by_trackerはfeatureを必要としているのでshape_estimationの出力にFeatureObjectsを追加

detection_by_trackerを使わないような場合には
shape_estimation -> trackingとなるので DynamicObjects型のoutputは継続して残しました

**Comment by tkimura4** [comment]
> 素直にconverterノード作っても良いかなという気もしますが、どう思います？

なるほど、そうしますか.

**Comment by tkimura4** [comment]
- shape_estimationの出力をDynamicObjects から DynamicObjectsWithFeatureArrayに変更
- dummy_perception_publisher -> shape_estimation -> trackerなど, 直接trackerと繋がる場合は, 間に一段converterを挟むように
- これに伴ってdummy_perception_publisherの出力タイプ変更が不要になったため削除

**Comment by taichiH** [comment]
@tkimura4 I think `detected_object_converter` is more appropriate node name. How do you think?

File: perception/util/converter/dynamic_object_converter/include/dynamic_object_converter/dynamic_object_converter.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2021 Autoware Foundation
```
Comment (yukkysaito) [inline perception/util/converter/dynamic_object_converter/include/dynamic_object_converter/dynamic_object_converter.hpp:1]: Still on TierIV, please.
pls others

File: perception/util/converter/dynamic_object_converter/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic -Werror)
+endif()
+
+### Dependencies
+find_package(ament_cmake_auto REQUIRED)
+ament_auto_find_build_dependencies()
+
+# Generate exe file
+ament_auto_add_library(dynamic_object_converter_node SHARED
```
Comment (yukkysaito) [inline perception/util/converter/dynamic_object_converter/CMakeLists.txt:19]: The word "dynamic object" is no longer used.
Candidate list.
- detected_object_feature_remover
- detected_object_converter

**Comment by yukkysaito** [comment]
TODO
- [ ] readme
- [ ] change dynamic_object_converter name
- [ ] license

File: perception/util/converter/dynamic_object_converter/include/dynamic_object_converter/dynamic_object_converter.hpp (hunk 1 above)
Comment (tkimura4) [reply perception/util/converter/dynamic_object_converter/include/dynamic_object_converter/dynamic_object_converter.hpp]: Fixed at a162d36

File: perception/util/converter/dynamic_object_converter/CMakeLists.txt (hunk 2 above)
Comment (tkimura4) [reply perception/util/converter/dynamic_object_converter/CMakeLists.txt]: -> detected_object_feature_remover
7f5994d

**Comment by tkimura4** [comment]
Readme: f46eccb

File: perception/object_recognition/detection/shape_estimation/src/node.hpp [hunk 3]
```diff
@@ -24,18 +24,18 @@
...
 class ShapeEstimationNode : public rclcpp::Node
 {
 private:
   // ros
-  rclcpp::Publisher<autoware_auto_perception_msgs::msg::DetectedObjects>::SharedPtr pub_;
-  rclcpp::Subscription<autoware_perception_msgs::msg::DetectedObjectsWithFeature>::SharedPtr sub_;
-  // bool use_map_correct_;
+  rclcpp::Publisher<DetectedObjectsWithFeature>::SharedPtr pub_;
+  rclcpp::Subscription<DetectedObjectsWithFeature>::SharedPtr sub_;
+  bool output_object_with_feature_;
```
Comment (yukkysaito) [inline perception/object_recognition/detection/shape_estimation/src/node.hpp:35]: 不要？

File: perception/util/converter/detected_object_feature_remover/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# detected_object_feature_remover
+
+## Purpose
+
+The `detected_object_feature_remover` is a package to convert topic-type from `DetectedObjectWithFeatureArray` to `DetectedObjects`.
+
+![traffic light roi visualization](./images/roi-visualization.png)
```
Comment (yukkysaito) [inline perception/util/converter/detected_object_feature_remover/README.md:7]: 不要？

File: perception/util/converter/detected_object_feature_remover/include/detected_object_feature_remover/detected_object_feature_remover.hpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+class DetectedObjectFeatureRemover : public rclcpp::Node
+{
+public:
+  explicit DetectedObjectFeatureRemover(const rclcpp::NodeOptions & node_options);
+
+private:
+  rclcpp::Subscription<DetectedObjectsWithFeature>::SharedPtr sub_;
+  rclcpp::Publisher<DetectedObjects>::SharedPtr pub_;
+  void objectCallback(const DetectedObjectsWithFeature::ConstSharedPtr input);
+  DetectedObjects convert(const DetectedObjectsWithFeature & objs_with_feature);
```
Comment (yukkysaito) [inline perception/util/converter/detected_object_feature_remover/include/detected_object_feature_remover/detected_object_feature_remover.hpp:37]: DetectedObjectsはそこそこ大きいので
```
void convert(const DetectedObjectsWithFeature & objs_with_feature, DetectedObjects & objs);
```
のほうが良いと思います

File: simulator/util/dummy_perception_publisher/launch/dummy_perception_publisher.launch.xml [hunk 6]
```diff
@@ -19,16 +19,15 @@
         <param name="detection_successful_rate" value="0.999" />
         <param name="enable_ray_tracing" value="true" />
         <param name="use_object_recognition" value="$(var use_object_recognition)" />
-        <param name="use_real_param" value="true" />
       </node>
       <include file="$(find-pkg-share shape_estimation)/launch/shape_estimation.launch.xml">
         <arg name="input/objects" value="/perception/object_recognition/detection/labeled_clusters" />
-        <arg name="output/objects" value="/perception/object_recognition/detection/objects" />
+        <arg name="output/objects" value="/perception/object_recognition/detection/objects_raw" />
```
Comment (yukkysaito) [inline simulator/util/dummy_perception_publisher/launch/dummy_perception_publisher.launch.xml:25]: rawというのは分かりづらいのでobjects_with_featureで良い気がします

Comment (tkimura4) [reply simulator/util/dummy_perception_publisher/launch/dummy_perception_publisher.launch.xml]: 132aa46

File: perception/util/converter/detected_object_feature_remover/include/detected_object_feature_remover/detected_object_feature_remover.hpp (hunk 5 above)
Comment (tkimura4) [reply perception/util/converter/detected_object_feature_remover/include/detected_object_feature_remover/detected_object_feature_remover.hpp]: 8bab554

File: perception/util/converter/detected_object_feature_remover/README.md (hunk 4 above)
Comment (tkimura4) [reply perception/util/converter/detected_object_feature_remover/README.md]: fd2a261

File: perception/object_recognition/detection/shape_estimation/src/node.hpp (hunk 3 above)
Comment (tkimura4) [reply perception/object_recognition/detection/shape_estimation/src/node.hpp]: efc8e14

**Review by yukkysaito — APPROVED** [review]

**Pull request merged**